internal/discharger/login.go

```diff
@@ -152,17 +156,76 @@ func (h *handler) RedirectLogin(p httprequest.Params, req *redirectLoginRequest)
 	if len(idps) == 0 {
 		idps = allIDPs
 	}
-	idpChoices := params.IDPChoice{IDPs: idps}
-	if p.Request.Header.Get("Accept") == "application/json" {
-		httprequest.WriteJSON(p.Response, http.StatusOK, idpChoices)
+	if req.Header.Get("Accept") == "application/json" {
+		httprequest.WriteJSON(w, http.StatusOK, params.IDPChoice{IDPs: idps})
 		return nil
 	}
-	if err := h.params.Template.ExecuteTemplate(p.Response, "authentication-required", idpChoices); err != nil {
+	type authenticationRequiredParams struct {
+		IDPs          []params.IDPChoiceDetails
+		Error         string
+		UseEmail      bool
+		ShowEmailLink bool
+		WithEmailURL  string
+	}
+	authParams := authenticationRequiredParams{
+		IDPs:          idps,
+		Error:         errorMessage,
+		UseEmail:      useEmail,
+		ShowEmailLink: domain == "" && !useEmail,
+		WithEmailURL:  h.params.Location + "/login-email?state=" + state,
+	}
+	if err := h.params.Template.ExecuteTemplate(w, "authentication-required", authParams); err != nil {
 		return errgo.Mask(err)
 	}
 	return nil
 }
 
+type emailLoginRequest struct {
+	httprequest.Route `httprequest:"GET /login-email"`
+
+	// State holds the state information generated in the /login-redirect
+	// handler.
+	State string `httprequest:"state,form"`
+}
+
+// EmailLogin starts a request to choose an identity provider using an
+// email address.
+func (h *handler) EmailLogin(p httprequest.Params, req *emailLoginRequest) error {
+	return h.authChoice(p.Response, p.Request, req.State, "", "", true)
+}
+
+type emailLoginSubmitRequest struct {
+	httprequest.Route `httprequest:"POST /login-email"`
+
+	// State holds the state information generated in the /login-redirect
+	// handler.
+	State string `httprequest:"state,form"`
+
+	// Email holds the email address to use to select the identity
+	// provider.
+	Email string `httprequest:"email,form"`
+}
+
+// EmailLoginSubmit attemtps to find an identity provider suitable for
+// use with the provided email address and either redirects to that
+// identity provider, or outputs the same email form with an error message.
+func (h *handler) EmailLoginSubmit(p httprequest.Params, req *emailLoginSubmitRequest) error {
+	for _, idp := range h.params.IdentityProviders {
+		type isForEmailAddrer interface {
```

Comment:
Is this a typo? Should it be `isForEmailAddr`?

Reply:
it's not a typo, but that doesn't mean it's a sensible name.